### ISSUE TITLE
Restore data after a Borg backup disk crash

### ISSUE THREAD

**Issue by prometheusAmb** [borgbackup/borg#7847]
## Have you checked borgbackup docs, FAQ, and open GitHub issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

Invalid

#### Operating system (distribution) and version.
Deepin Linux 20.9 Community Edition
#### Hardware / network configuration, and filesystems used.
HP Envy 13.3 inch (HP ENVY x360 Convertible 13-ay1xxx **(54B73PA#ACJ)
**
#### How much data is handled by borg?
Approximately 830 GB
## Describe the problem you're observing.
I have a backup of my data (entire home drive) on an external Western Digital USB hard drive which is around 7 months old. It has been working fine for the last several months. However, day before yesterday, the drive developed a problem. Its contents could no longer be displayed in the file manager. File manager said the drive needs to be formatted.

I did not format it but I used testdisk to find the data. I found the borg backup folder in testdisk and restored it to a new USB drive along with all configs etc. Borg did not recognize it as a valid backup.

I created a new borg repository and copied over the data to the new repository but no luck. I have the data but cannot access it (so it seems)

**Comment by ThomasWaldmann** [comment]
You need the repo config from the original repo directory.

In case that doesn't help: I need more precise information, like what commands you used, error msgs you get, borg version, ...

**Comment by prometheusAmb** [comment]
I have copied the repo config from the original to the new backup destination.

Running check --repair on repository only was OK.

But running check --repair on archives only is spewing out messages about corrupted chunk and "Encryption envelope checksum mismatch"

**Comment by prometheusAmb** [comment]
> You need the repo config from the original repo directory.
> 
> In case that doesn't help: I need more precise information, like what commands you used, error msgs you get, borg version, ...

Thanks for your prompt response. 

I used the command "borg list" on the repository first. It showed me an invalid repository error

I then created a new repository and added the data files from the old repository along with the configs and ran the check --repair commands above with the results given above.

Thanks again for your response

**Comment by ThomasWaldmann** [comment]
> Running check --repair on repository only was OK.

That was doing a crc32 check on all repo objects (and a repo index rebuild).

> But running check --repair on archives only is spewing out messages about
> corrupted chunk and "Encryption envelope checksum mismatch"

That shows corruption/tampering in the repository: the stronger cryptographic checks found corrupted/tampered data which the crc32 check did not find - that's rather unusual.

Is your hw working reliably?

**Comment by prometheusAmb** [comment]
Thanks again for your very prompt response. The question is what should I do? The backup drive developed a problem. I had to use testdisk to get the data out of the drive. Testdisk gave me all the data files and configs contained in the backup repository. But borg list on the repo did not work. So I did check --repair. 

If the stronger cryptographic checks find inconsistencies does that mean that the backup is lost?

**Comment by ThomasWaldmann** [comment]
- copy repo_dir/data from old to new
- copy repo_dir/config from old to new
- run `borg check --repair -v --progress ...`

borg will try to repair / rescue all that is not corrupted.

how good the result will be depends on the amount of corruption.

make sure your hw is working correctly, esp. RAM and storage before doing the repair.

**Comment by prometheusAmb** [comment]
> * copy repo_dir/data from old to new
> 
>     * copy repo_dir/config from old to new
> 
>     * run `borg check --repair -v --progress ...` on the new disk
> 
> 
> borg will try to repair / rescue all that is not corrupted.
> 
> how good the result will be depends on the amount of corruption.
> 
> make sure your hw is working correctly, esp. RAM and storage before doing the repair.
> 
> also, do not write `did not work`. I need the full commands and their output to see what exactly did not work.

Many thanks for your response outlining all the steps. Really appreciated. And yes, as you have pointed out, I should have included the full response but couldn't in the business of trying to get my data recovered

As it happens, the backup is now recovered. I was able to mount the backup on the computer and copy the files I needed to. What I did to achieve this was to copy the config files from the old (corrupted backup retrieved by testdisk) once again to the new backup destination and did a check:

borg check --repair -v -p --repository-only new-repo. 

Borg checked and was happy with the check.

Then I was able to list the archives by:

borg list new-repo

Contents of the repo were listed. I chose the one that I wanted to restore from and mounted it using borg mount with success. 

So the backup is now recovered (thanks to testdisk) and securely located in a new destination drive.

Thanks once again for all the help
